// happy-dom/lib/vm-context/WebpackScriptCompiler.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var VM = __importStar(require("vm"));
var memory_fs_1 = __importDefault(require("memory-fs"));
var webpack_1 = __importDefault(require("webpack"));
var DEFAULT_CONFIG = {
    mode: 'production',
    devtool: false,
    optimization: {
        minimize: false
    }
};
/**
 * This class is used for handling a web component server dom.
 */
var WebpackScriptCompiler = /** @class */ (function () {
    function WebpackScriptCompiler() {
    }
    /**
     * Setup of scripts.
     *
     * @param {IWebpackConfig} config Webpack config.
     * @return {Promise<VM.Script>} Promise.
     */
    WebpackScriptCompiler.prototype.compile = function (config) {
        return new Promise(function (resolve, reject) {
            var compiler = webpack_1.default(Object.assign({}, config, DEFAULT_CONFIG, {
                output: {
                    filename: 'output.js',
                    path: '/'
                }
            }));
            compiler.outputFileSystem = new memory_fs_1.default();
            compiler.run(function (error, stats) {
                if (error) {
                    reject(error);
                }
                else {
                    if (stats.hasErrors()) {
                        reject(new Error('Errors occured during compilation. Error message: ' + stats.toJson('errors-only').errors.join('\n')));
                    }
                    else {
                        var result = compiler.outputFileSystem.data['output.js'].toString();
                        resolve(new VM.Script(result));
                    }
                }
            });
        });
    };
    return WebpackScriptCompiler;
}());
exports.default = WebpackScriptCompiler;
