// happy-dom/lib/nodes/basic/element/Element.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Node_1 = __importDefault(require("../node/Node"));
var TextNode_1 = __importDefault(require("../text-node/TextNode"));
var NodeType_1 = __importDefault(require("../node/NodeType"));
var ShadowRoot_1 = __importDefault(require("../shadow-root/ShadowRoot"));
var Attribute_1 = __importDefault(require("./Attribute"));
var DOMRect_1 = __importDefault(require("./DOMRect"));
var Range_1 = __importDefault(require("./Range"));
var HTMLParser_1 = __importDefault(require("../../../html-parser/HTMLParser"));
var he_1 = require("he");
var ClassList_1 = __importDefault(require("./ClassList"));
var QuerySelector_1 = __importDefault(require("../../../query-selector/QuerySelector"));
var HTMLRenderer_1 = __importDefault(require("../../../html-renderer/HTMLRenderer"));
var MutationRecord_1 = __importDefault(require("../../../mutation-observer/MutationRecord"));
var MutationType_1 = __importDefault(require("../../../mutation-observer/MutationType"));
var ATTRIBUTE_REGEXP = /([^\s=]+)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|(\S+)))/gi;
/**
 * Element.
 */
var Element = /** @class */ (function (_super) {
    __extends(Element, _super);
    function Element() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.tagName = null;
        _this.nodeType = NodeType_1.default.ELEMENT_NODE;
        _this.shadowRoot = null;
        _this.classList = new ClassList_1.default(_this);
        _this.scrollTop = 0;
        _this.scrollLeft = 0;
        _this._attributesMap = {};
        _this._attributePropertyMap = null;
        _this._useCaseSensitiveAttributes = false;
        return _this;
    }
    Object.defineProperty(Element.prototype, "id", {
        /**
         * Returns ID.
         *
         * @return ID.
         */
        get: function () {
            return this.getAttribute('id');
        },
        /**
         * Sets ID.
         *
         * @param id ID.
         * @return HTML.
         */
        set: function (id) {
            this.setAttribute('id', id);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "className", {
        /**
         * Returns class name.
         *
         * @return Class name.
         */
        get: function () {
            return this.getAttribute('class');
        },
        /**
         * Sets class name.
         *
         * @param className Class name.
         * @return Class name.
         */
        set: function (className) {
            this.setAttribute('class', className);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "children", {
        /**
         * Returns children.
         *
         * @returns Children.
         */
        get: function () {
            return this.childNodes.filter(function (childNode) { return childNode instanceof Element; });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "nodeName", {
        /**
         * Node name.
         *
         * @return Node name.
         */
        get: function () {
            return this.tagName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "textContent", {
        /**
         * Get text value of children.
         *
         * @return Text content.
         */
        get: function () {
            var result = '';
            for (var _i = 0, _a = this.childNodes; _i < _a.length; _i++) {
                var childNode = _a[_i];
                if (childNode instanceof Element || childNode instanceof TextNode_1.default) {
                    result += childNode.textContent;
                }
            }
            return result;
        },
        /**
         * Sets text content.
         *
         * @param textContent Text content.
         */
        set: function (textContent) {
            for (var _i = 0, _a = this.childNodes.slice(); _i < _a.length; _i++) {
                var child = _a[_i];
                this.removeChild(child);
            }
            this.appendChild(this.ownerDocument.createTextNode(textContent));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "innerHTML", {
        /**
         * Returns inner HTML.
         *
         * @return HTML.
         */
        get: function () {
            return HTMLRenderer_1.default.getInnerHTML(this);
        },
        /**
         * Sets inner HTML.
         *
         * @param html HTML.
         */
        set: function (html) {
            var root = HTMLParser_1.default.parse(this.ownerDocument, html);
            for (var _i = 0, _a = this.childNodes.slice(); _i < _a.length; _i++) {
                var child = _a[_i];
                this.removeChild(child);
            }
            for (var _b = 0, _c = root.childNodes.slice(); _b < _c.length; _b++) {
                var child = _c[_b];
                this.appendChild(child);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "outerHTML", {
        /**
         * Returns outer HTML.
         *
         * @return HTML.
         */
        get: function () {
            return HTMLRenderer_1.default.getOuterHTML(this);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Element.prototype, "attributes", {
        /**
         * Returns attributes.
         *
         * @returns Attributes.
         */
        get: function () {
            var names = Object.keys(this._attributesMap);
            var attributes = { length: names.length };
            for (var i = 0, max = names.length; i < max; i++) {
                var name = names[i];
                var attribute = new Attribute_1.default();
                attribute.name = name;
                attribute.value = this._attributesMap[name];
                attributes[name] = attribute;
                attributes[i] = attribute;
            }
            return attributes;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Sets an attribute.
     *
     * @param name Name.
     * @param value Value.
     */
    Element.prototype.setAttribute = function (name, value) {
        var lowerName = this._useCaseSensitiveAttributes ? name : name.toLowerCase();
        var oldValue = this._attributesMap[lowerName] !== undefined ? this._attributesMap[lowerName] : null;
        this._attributesMap[lowerName] = String(value);
        this._setAttributeProperty(lowerName, this._attributesMap[lowerName]);
        if (this.attributeChangedCallback) {
            this.attributeChangedCallback(name, oldValue, value);
        }
        // MutationObserver
        if (this._observers.length > 0) {
            for (var _i = 0, _a = this._observers; _i < _a.length; _i++) {
                var observer = _a[_i];
                if (observer.options.attributes &&
                    (!observer.options.attributeFilter || observer.options.attributeFilter.includes(lowerName))) {
                    var record = new MutationRecord_1.default();
                    record.type = MutationType_1.default.attributes;
                    record.attributeName = lowerName;
                    record.oldValue = observer.options.attributeOldValue ? oldValue : null;
                    observer.callback([record]);
                }
            }
        }
    };
    /**
     * Returns attribute value.
     *
     * @param name Name.
     */
    Element.prototype.getAttribute = function (name) {
        var lowerName = this._useCaseSensitiveAttributes ? name : name.toLowerCase();
        return this.hasAttribute(name) ? this._attributesMap[lowerName] : null;
    };
    /**
     * Returns a boolean value indicating whether the specified element has the attribute or not.
     *
     * @param name Attribute name.
     * @returns True if attribute exists, false otherwise.
     */
    Element.prototype.hasAttribute = function (name) {
        var lowerName = this._useCaseSensitiveAttributes ? name : name.toLowerCase();
        return this._attributesMap[lowerName] !== undefined;
    };
    /**
     * Returns "true" if the element has attributes.
     *
     * @return "true" if the element has attributes.
     */
    Element.prototype.hasAttributes = function () {
        return Object.keys(this._attributesMap).length > 0;
    };
    /**
     * Removes an attribute.
     *
     * @param name Name.
     */
    Element.prototype.removeAttribute = function (name) {
        var lowerName = this._useCaseSensitiveAttributes ? name : name.toLowerCase();
        var oldValue = this._attributesMap[lowerName] !== undefined ? this._attributesMap[lowerName] : null;
        delete this._attributesMap[lowerName];
        this._removeAttributeProperty(name);
        if (this.attributeChangedCallback) {
            this.attributeChangedCallback(name, oldValue, null);
        }
        // MutationObserver
        if (this._observers.length > 0) {
            for (var _i = 0, _a = this._observers; _i < _a.length; _i++) {
                var observer = _a[_i];
                if (observer.options.attributes &&
                    (!observer.options.attributeFilter || observer.options.attributeFilter.includes(lowerName))) {
                    var record = new MutationRecord_1.default();
                    record.type = MutationType_1.default.attributes;
                    record.attributeName = lowerName;
                    record.oldValue = observer.options.attributeOldValue ? oldValue : null;
                    observer.callback([record]);
                }
            }
        }
    };
    /**
     * Sets raw attributes.
     *
     * @param rawAttributes Raw attributes.
     */
    Element.prototype._setRawAttributes = function (rawAttributes) {
        rawAttributes = rawAttributes.trim();
        if (rawAttributes) {
            var regExp = new RegExp(ATTRIBUTE_REGEXP, 'gi');
            var match = void 0;
            // Attributes with value
            while ((match = regExp.exec(rawAttributes))) {
                var name = this._useCaseSensitiveAttributes ? match[1] : match[1].toLowerCase();
                if (name) {
                    this._attributesMap[name] = he_1.decode(match[2] || match[3] || match[4] || '');
                    this._setAttributeProperty(name, this._attributesMap[name]);
                }
            }
            // Attributes with no value
            for (var _i = 0, _a = rawAttributes
                .replace(ATTRIBUTE_REGEXP, '')
                .trim()
                .split(' '); _i < _a.length; _i++) {
                var name = _a[_i];
                var lowerName = this._useCaseSensitiveAttributes ? name.trim() : name.trim().toLowerCase();
                if (lowerName) {
                    this._attributesMap[lowerName] = '';
                    this._setAttributeProperty(lowerName, this._attributesMap[lowerName]);
                }
            }
        }
    };
    /**
     * Returns raw attributes.
     *
     * @returns {string} Raw attributes.
     */
    Element.prototype._getRawAttributes = function () {
        var attributes = [];
        for (var _i = 0, _a = Object.keys(this._attributesMap); _i < _a.length; _i++) {
            var name = _a[_i];
            if (this._attributesMap[name]) {
                attributes.push(name + '="' + he_1.encode(this._attributesMap[name]) + '"');
            }
            else if (this._attributesMap[name] !== undefined) {
                attributes.push(name);
            }
        }
        return attributes.join(' ');
    };
    /**
     * Attaches a shadow root.
     *
     * @param _shadowRootInit Shadow root init.
     * @returns Shadow root.
     */
    Element.prototype.attachShadow = function (_shadowRootInit) {
        if (this.shadowRoot) {
            throw new Error('Shadow root has already been attached.');
        }
        this.shadowRoot = new ShadowRoot_1.default();
        this.shadowRoot.ownerDocument = this.ownerDocument;
        this.shadowRoot.isConnected = this.isConnected;
        return this.shadowRoot;
    };
    /**
     * Scrolls to a particular set of coordinates in the document.
     *
     * @note This method has not been implemented. It is just here for compatibility.
     */
    Element.prototype.scrollTo = function () { };
    /**
     * Converts to string.
     *
     * @return String.
     */
    Element.prototype.toString = function () {
        return this.outerHTML;
    };
    /**
     * Returns the size of an element and its position relative to the viewport.
     *
     * @returns DOM rect.
     */
    Element.prototype.getBoundingClientRect = function () {
        return new DOMRect_1.default();
    };
    /**
     * Returns a range.
     *
     * @returns Range.
     */
    Element.prototype.createTextRange = function () {
        return new Range_1.default();
    };
    /**
     * Query CSS selector to find matching nodes.
     *
     * @param selector CSS selector.
     * @returns Matching elements.
     */
    Element.prototype.querySelectorAll = function (selector) {
        return QuerySelector_1.default.querySelectorAll(this, selector);
    };
    /**
     * Query CSS Selector to find matching node.
     *
     * @param selector CSS selector.
     * @return Matching node.
     */
    Element.prototype.querySelector = function (selector) {
        return QuerySelector_1.default.querySelector(this, selector);
    };
    /**
     * Returns an elements by tag name.
     *
     * @param tagName Tag name.
     * @returns Matching nodes.
     */
    Element.prototype.getElementsByTagName = function (tagName) {
        return this.querySelectorAll(tagName);
    };
    /**
     * Returns an elements by class name.
     *
     * @param className Tag name.
     * @returns Matching nodes.
     */
    Element.prototype.getElementsByClassName = function (className) {
        return this.querySelectorAll('.' + className.split(' ').join('.'));
    };
    /**
     * Sets a property when setting an attribute.
     *
     * @param name Name.
     * @param value Value.
     */
    Element.prototype._setAttributeProperty = function (name, value) {
        if (name === 'style' && this['style']) {
            for (var _i = 0, _a = value.split(';'); _i < _a.length; _i++) {
                var part = _a[_i];
                var _b = part.split(':'), key = _b[0], value_1 = _b[1];
                this['style'][key.trim()] = value_1.trim();
            }
        }
        else {
            var property = this._getPropertyNameFromAttribute(name);
            if (property) {
                if (this[property] === '') {
                    this[property] = value;
                }
                else if (typeof this[property] === 'boolean') {
                    this[property] = true;
                }
                else if (typeof this[property] === 'number') {
                    this[property] = parseFloat(value);
                }
            }
        }
    };
    /**
     * Sets boolean properties to false if a matching attribute is removed.
     *
     * @param name Name.
     */
    Element.prototype._removeAttributeProperty = function (name) {
        var property = this._getPropertyNameFromAttribute(name);
        if (property && typeof this[property] === 'boolean') {
            this[property] = false;
        }
    };
    /**
     * Returns the property name based on an attribute name.
     * This will be used for setting attribute values as properties on the element.
     *
     * @param name Attribute name.
     * @returns Property name.
     */
    Element.prototype._getPropertyNameFromAttribute = function (name) {
        if (!this._attributePropertyMap) {
            this._attributePropertyMap = {};
            for (var _i = 0, _a = Object.keys(this); _i < _a.length; _i++) {
                var key = _a[_i];
                this._attributePropertyMap[key.toLowerCase()] = key;
            }
        }
        return this._attributePropertyMap[name] || null;
    };
    return Element;
}(Node_1.default));
exports.default = Element;
