// happy-dom/lib/nodes/basic-types/html-element/HTMLElement.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Element_1 = __importDefault(require("../element/Element"));
var Event_1 = __importDefault(require("../../../event/Event"));
/**
 * HTMLElement.
 */
var HTMLElement = /** @class */ (function (_super) {
    __extends(HTMLElement, _super);
    function HTMLElement() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.style = {};
        _this.tabIndex = 0;
        _this.offsetHeight = 0;
        _this.offsetWidth = 0;
        _this.offsetLeft = 0;
        _this.offsetTop = 0;
        _this.clientHeight = 0;
        _this.clientWidth = 0;
        return _this;
    }
    Object.defineProperty(HTMLElement.prototype, "innerText", {
        /**
         * Returns inner text.
         *
         * @return {string} Text.
         */
        get: function () {
            return this.textContent;
        },
        /**
         * Sets inner text.
         *
         * @param {string} text Text.
         */
        set: function (text) {
            this.textContent = text;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Triggers a click event.
     */
    HTMLElement.prototype.click = function () {
        this.dispatchEvent(new Event_1.default('click'));
    };
    /**
     * Triggers a blur event.
     */
    HTMLElement.prototype.blur = function () {
        this.dispatchEvent(new Event_1.default('blur'));
    };
    /**
     * Triggers a focus event.
     */
    HTMLElement.prototype.focus = function () {
        this.dispatchEvent(new Event_1.default('focus'));
    };
    /**
     * Sets an attribute.MouseEvent
     *
     * @override
     * @param {string} name Name.
     * @param {string} value Value.
     */
    HTMLElement.prototype.setAttribute = function (name, value) {
        var lowerName = this._useCaseSensitiveAttributes ? name : name.toLowerCase();
        _super.prototype.setAttribute.call(this, lowerName, value);
        var observedPropertyAttributes = this.constructor._observedPropertyAttributes;
        var observedAttributes = Object.keys(observedPropertyAttributes);
        if (value !== null && value !== undefined && observedAttributes.includes(lowerName)) {
            var property = observedPropertyAttributes[lowerName];
            this[property] = typeof this[property] === 'boolean' ? value !== null : String(value);
        }
    };
    /**
     * Sets raw attributes.
     *
     * @override
     * @param {string} rawAttributes Raw attributes.
     */
    HTMLElement.prototype._setRawAttributes = function (rawAttributes) {
        _super.prototype._setRawAttributes.call(this, rawAttributes);
        if (rawAttributes.trim()) {
            this._defineInitialProperties();
        }
    };
    /**
     * Defines initial properties.
     */
    HTMLElement.prototype._defineInitialProperties = function () {
        var observedPropertyAttributes = this.constructor._observedPropertyAttributes;
        for (var _i = 0, _a = Object.keys(observedPropertyAttributes); _i < _a.length; _i++) {
            var name = _a[_i];
            var attribute = this._attributesMap[name];
            if (attribute !== null && attribute !== undefined) {
                var property = observedPropertyAttributes[name];
                switch (typeof this[property]) {
                    case 'boolean':
                        this[property] = true;
                    case 'number':
                        this[property] = parseFloat(attribute);
                    default:
                        this[property] = attribute;
                }
            }
        }
    };
    HTMLElement._observedPropertyAttributes = { tabindex: 'tabIndex' };
    return HTMLElement;
}(Element_1.default));
exports.default = HTMLElement;
