// happy-dom/lib/event/Event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Event = /** @class */ (function () {
    /**
     * Constructor.
     *
     * @param {string} type Event type.
     * @param {IEventInit} eventInit Event init.
     */
    function Event(type, eventInit) {
        if (eventInit === void 0) { eventInit = null; }
        this.bubbles = false;
        this.cancelable = false;
        this.composed = false;
        this.currentTarget = null;
        this.target = null;
        this.defaultPrevented = false;
        this._immediatePropagationStopped = false;
        this._propagationStopped = false;
        this.type = null;
        this.type = type;
        this.bubbles = eventInit && eventInit.bubbles ? true : false;
        this.cancelable = eventInit && eventInit.cancelable ? true : false;
        this.composed = eventInit && eventInit.composed ? true : false;
    }
    /**
     * Init event.
     *
     * @legacy
     */
    Event.prototype.initEvent = function () {
        this.defaultPrevented = true;
    };
    /**
     * Prevents default.
     */
    Event.prototype.preventDefault = function () {
        this.defaultPrevented = true;
    };
    /**
     * Stops immediate propagation.
     */
    Event.prototype.stopImmediatePropagation = function () {
        this._immediatePropagationStopped = true;
    };
    /**
     * Stops propagation.
     */
    Event.prototype.stopPropagation = function () {
        this._propagationStopped = true;
    };
    return Event;
}());
exports.default = Event;
