// happy-dom/lib/html-renderer/shadow-root/css/ScopedCSSCache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ABC = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
/**
 * CSS Cache.
 */
var ScopedCSSCache = /** @class */ (function () {
    function ScopedCSSCache() {
        this.original = [];
        this.scoped = [];
    }
    /**
     * Returns the scoped CSS.
     *
     * @return {string[]} Scoped style.
     */
    ScopedCSSCache.prototype.getAllScoped = function () {
        return this.scoped.concat();
    };
    /**
     * Returns scoped CSS.
     *
     * @param {string} original Original CSS.
     * @return {string} Cached scoped CSS.
     */
    ScopedCSSCache.prototype.getScoped = function (original) {
        var index = this.original.indexOf(original);
        if (index !== -1) {
            return this.scoped[index];
        }
        return null;
    };
    /**
     * Returns scoped CSS.
     *
     * @param {string} original Original CSS.
     * @param {string} scopedCSS Scoped CSS.
     */
    ScopedCSSCache.prototype.setScoped = function (original, scopedCSS) {
        var index = this.original.indexOf(original);
        if (index !== -1) {
            this.original[index] = original;
            this.scoped[index] = scopedCSS;
        }
        else {
            this.original.push(original);
            this.scoped.push(scopedCSS);
        }
    };
    /**
     * Removes a cache entry.
     *
     * @param {string} original Original CSS.
     */
    ScopedCSSCache.prototype.removeScoped = function (original) {
        var index = this.original.indexOf(original);
        if (index !== -1) {
            this.original.splice(index, 1);
            this.scoped.splice(index, 1);
        }
    };
    /**
     * Returns scope ID.
     *
     * @param {string} original Original CSS.
     * @return string Scope ID.
     */
    ScopedCSSCache.prototype.getScopeID = function (original) {
        var index = this.original.indexOf(original);
        if (index !== -1) {
            return this.getIdByIndex(index);
        }
        return this.getIdByIndex(this.original.length);
    };
    /**
     * Clears cache.
     */
    ScopedCSSCache.prototype.clear = function () {
        this.original = [];
        this.scoped = [];
    };
    /**
     * Returns an unique ID.
     *
     * @param {number} index Index.
     * @return {string} ID.
     */
    ScopedCSSCache.prototype.getIdByIndex = function (index) {
        return ABC[index] !== undefined ? ABC[index] : 'a' + index;
    };
    return ScopedCSSCache;
}());
exports.default = ScopedCSSCache;
