// happy-dom/lib/location/URL.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var URL_REGEXP = /(https?:)\/\/([-a-zA-Z0-9@:%._\+~#=]{2,256}[a-z]{2,6})(:[0-9]*)?([-a-zA-Z0-9@:%_\+.~c&//=]*)(\?[^#]*)?(#.*)?/;
var URL = /** @class */ (function () {
    /**
     * Constructor.
     *
     * @param {string} url URL.
     */
    function URL(url) {
        this.protocol = '';
        this.hostname = '';
        this.port = '';
        this.pathname = '';
        this.search = '';
        this.hash = '';
        this.username = '';
        this.password = '';
        if (url) {
            this.parse(url);
        }
    }
    /**
     * Returns the entire URL as a string.
     */
    URL.prototype.toString = function () {
        return this.href;
    };
    /**
     * Parses an URL.
     *
     * @param {string} url URL.
     */
    URL.prototype.parse = function (url) {
        var match = url.match(URL_REGEXP);
        if (!match) {
            throw new Error('Failed to parse URL "' + url + '". URL is in an invalid format.');
        }
        var hostnamePart = match[2] ? match[2].split('@') : '';
        var credentialsPart = hostnamePart.length > 1 ? hostnamePart[0].split(':') : null;
        this.protocol = match[1] || '';
        this.hostname = hostnamePart.length > 1 ? hostnamePart[1] : hostnamePart[0];
        this.port = match[3] || '';
        this.pathname = match[4] || '';
        this.search = match[5] || '';
        this.hash = match[6] || '';
        this.username = credentialsPart ? credentialsPart[0] : '';
        this.password = credentialsPart ? credentialsPart[1] : '';
    };
    Object.defineProperty(URL.prototype, "href", {
        /**
         * Returns the entire URL as a string.
         *
         * @return {string} Href.
         */
        get: function () {
            var credentials = this.username ? this.username + ":" + this.password + "@" : '';
            return this.protocol + '//' + credentials + this.host + this.pathname + this.search + this.hash;
        },
        /**
         * Sets the href.
         *
         * @param {string} url URL.
         */
        set: function (url) {
            this.parse(url);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(URL.prototype, "origin", {
        /**
         * Returns the origin.
         *
         * @return {string} HREF.
         */
        get: function () {
            return this.protocol + '//' + this.host;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(URL.prototype, "host", {
        /**
         * Returns the entire URL as a string.
         *
         * @return {string} Host.
         */
        get: function () {
            return this.hostname + this.port;
        },
        enumerable: true,
        configurable: true
    });
    return URL;
}());
exports.default = URL;
