// happy-dom/lib/html-config/Elements.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var HTMLTemplateElement_1 = __importDefault(require("../nodes/elements/template/HTMLTemplateElement"));
var HTMLFormElement_1 = __importDefault(require("../nodes/elements/form/HTMLFormElement"));
var HTMLInputElement_1 = __importDefault(require("../nodes/elements/input/HTMLInputElement"));
var HTMLTextAreaElement_1 = __importDefault(require("../nodes/elements/text-area/HTMLTextAreaElement"));
var SVGSVGElement_1 = __importDefault(require("../nodes/elements/svg/SVGSVGElement"));
var SVGElement_1 = __importDefault(require("../nodes/elements/svg/SVGElement"));
exports.default = {
    template: HTMLTemplateElement_1.default,
    form: HTMLFormElement_1.default,
    input: HTMLInputElement_1.default,
    textarea: HTMLTextAreaElement_1.default,
    svg: SVGSVGElement_1.default,
    circle: SVGElement_1.default,
    ellipse: SVGElement_1.default,
    line: SVGElement_1.default,
    path: SVGElement_1.default,
    polygon: SVGElement_1.default,
    polyline: SVGElement_1.default,
    rect: SVGElement_1.default,
    stop: SVGElement_1.default,
    use: SVGElement_1.default
};
