// happy-dom/lib/nodes/elements/template/HTMLTemplateElement.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var HTMLElement_1 = __importDefault(require("../../basic-types/html-element/HTMLElement"));
/**
 * HTMLTemplateElement.
 */
var HTMLTemplateElement = /** @class */ (function (_super) {
    __extends(HTMLTemplateElement, _super);
    function HTMLTemplateElement() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.contentElement = null;
        return _this;
    }
    Object.defineProperty(HTMLTemplateElement.prototype, "content", {
        /**
         * Returns the content.
         *
         * @return {DocumentFragment} Content.
         */
        get: function () {
            if (!this.contentElement) {
                this.contentElement = this.ownerDocument.createDocumentFragment();
            }
            return this.contentElement;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HTMLTemplateElement.prototype, "previousSibling", {
        /**
         * Previous sibling.
         *
         * @return {Node} Node.
         */
        get: function () {
            return this.content.previousSibling;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HTMLTemplateElement.prototype, "nextSibling", {
        /**
         * Next sibling.
         *
         * @return {Node} Node.
         */
        get: function () {
            return this.content.nextSibling;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HTMLTemplateElement.prototype, "firstChild", {
        /**
         * First child.
         *
         * @return {Node} Node.
         */
        get: function () {
            return this.content.firstChild;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(HTMLTemplateElement.prototype, "lastChild", {
        /**
         * Last child.
         *
         * @return {Node} Node.
         */
        get: function () {
            return this.content.lastChild;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Append a child node to childNodes.
     *
     * @param  {Node} node Node to append.
     * @return {Node} Appended node.
     */
    HTMLTemplateElement.prototype.appendChild = function (node) {
        return this.content.appendChild(node);
    };
    /**
     * Remove Child element from childNodes array.
     *
     * @param {Node} node Node to remove
     */
    HTMLTemplateElement.prototype.removeChild = function (node) {
        return this.content.removeChild(node);
    };
    /**
     * Inserts a node before another.
     *
     * @param {Node} newNode Node to insert.
     * @param {Node} referenceNode Node to insert before.
     * @return {Node} Inserted node.
     */
    HTMLTemplateElement.prototype.insertBefore = function (newNode, referenceNode) {
        return this.content.insertBefore(newNode, referenceNode);
    };
    /**
     * Replaces a node with another.
     *
     * @param {Node} newChild New child.
     * @param {Node} oldChild Old child.
     * @return {Node} Replaced node.
     */
    HTMLTemplateElement.prototype.replaceChild = function (newChild, oldChild) {
        return this.content.replaceChild(newChild, oldChild);
    };
    return HTMLTemplateElement;
}(HTMLElement_1.default));
exports.default = HTMLTemplateElement;
